File: cpp/benchmark.cpp
#include <ctime>
#include <iostream>
#include <regex>
#include <string>

#include "regex.h"

using regex::Regex;
using std::cout;
using std::endl;
using std::string;

template <typename F>
double benchmark_micros(const F& f, int num_times = 10000) {
  auto start = std::clock();
  for (int i = 0; i < num_times; i++) {
    f();
  }
  auto end = std::clock();
  return ((end - start) / num_times) / (CLOCKS_PER_SEC / 1e6);
}

int main() {
  Regex r("(foo(ba)?)*(bar)+");
  std::regex re("(foo(ba)?)*(bar)+");
  const string test = "foofoobabarbar";
  cout << "Custom implementation matches: " << (r.match(test) ? "yes" : "no")
       << endl;
  cout << "Stdlib implementation matches: "
       << (std::regex_match(test, re) ? "yes" : "no") << endl;
  cout << "Custom implementation: "
       << benchmark_micros([&]() { r.match("foofoobabarbar"); })
       << " microseconds." << endl;
  cout << "Stdlib implementation: "
       << benchmark_micros([&]() { std::regex_match(test, re); })
       << " microseconds." << endl;
  return 0;
}

File: python/benchmark.py
import regex
import re
import time


def benchmark(f, num_times):
    start = time.time()
    for i in xrange(num_times):
        f()
    end = time.time()
    return (end - start) / float(num_times)


def benchmark_micros(f, precision=3):
    elapsed = benchmark(f, 100000)
    print "{{:.{precision}}} microseconds".format(precision=precision).format(elapsed * 1e6)


def main():
    pattern = '(foo(ba)?)*(bar)+'
    test = 'foofoobabarbar'

    r = regex.Regex(pattern)
    r2 = re.compile(pattern)

    print r.match(test)
    m = r2.match(test)
    print m is not None and m.group(0) == test

    benchmark_micros(lambda: r.match(test))
    benchmark_micros(lambda: r2.match(test))

if __name__ == "__main__":
    main()
